src/basic/btrfs-util.c

```diff
@@ -1689,8 +1689,7 @@ int btrfs_subvol_snapshot_fd(int old_fd, const char *new_path, BtrfsSnapshotFlag
                 r = btrfs_subvol_make(new_path);
                 if (r == -ENOTTY && (flags & BTRFS_SNAPSHOT_FALLBACK_DIRECTORY)) {
                         /* If the destination doesn't support subvolumes, then use a plain directory, if that's requested. */
-                        if (mkdir(new_path, 0755) < 0)
-                                return r;
```

Comment:
so the existing code returns the original error ENOTTY if the fallback failed. This is what trips up the logic here. As the caller of this API expects to get EEXIST if the destination already exists, and not ENOTTY. Hence, the right fix should be as easy as changing `return r` → `return -errno`, so that the error of the `mkdir()` is propagated, and not the original btrfs ioctl error.